### PR TITLE
[master] OPENJPA-2767: Incomplete ValueMapDiscriminatorStrategy cache and Meta…

### PR DESCRIPTION
**Pull request by dazey3**
https://issues.apache.org/jira/browse/OPENJPA-2767

Signed-off-by: Will Dazey <dazeydev.3@gmail.com>

### CODE REVIEW

File: openjpa-kernel/src/main/java/org/apache/openjpa/meta/MetaDataRepository.java [hunk 1]
```diff
@@ -1625,19 +1626,20 @@ private void loadRegisteredClassMetaData(ClassLoader envLoader) {
...
     /**
-     * Updates our datastructures with the latest registered classes.
+     * Updates our data structures with the latest registered classes.
+     * 
+     * This method is synchronized to make sure that all data structures are fully updated
+     *  before other threads attempt to call this method
      */
-    Class<?>[] processRegisteredClasses(ClassLoader envLoader) {
-        if (_registered.isEmpty())
+    synchronized Class<?>[] processRegisteredClasses(ClassLoader envLoader) {
```
Comment (ilgrosso) [inline openjpa-kernel/src/main/java/org/apache/openjpa/meta/MetaDataRepository.java]: Any chance to avoid synchronization at method level?

Comment (dazey3) [reply openjpa-kernel/src/main/java/org/apache/openjpa/meta/MetaDataRepository.java]: I had written a possible implementation using read/write locks similar to my ValueMapDiscriminator impl, but so much of the code ended up being write locked that it almost seemed pointless to have that fine grained locking. Almost all the time, `_registered` is empty anyway, so little to no blocking will occur.
The issue really ends up being that all of processRegisteredClasses needs to block until it's done processing all of the classes out of the temporary array. I realize the temp array was added to reduce blocking on the `_registered` array in the first place, but that only opens a window for other threads to start initializing other caches WHILE the MetaDataRepository is still processing and building. If the ORM is sufficiently large, I was seeing this processing taking up to 700ms. During that time, reads are potentially dirty from `_subs`.
That being said, if there is a strong performance concern from synchonizing the method (something that may not be as troubling nowadays with newer jdks), I can re-investigate implementing locks. This change was just infinitely cleaner.

Comment (ilgrosso) [reply openjpa-kernel/src/main/java/org/apache/openjpa/meta/MetaDataRepository.java]: @dazey3 I see: given what you say, method synchronization could be acceptable then - I was just wondering why...

Comment (dazey3) [reply openjpa-kernel/src/main/java/org/apache/openjpa/meta/MetaDataRepository.java]: @ilgrosso I understand the concern and I appreciate all the review comments I can get! Thank you for taking a look.

**Pull request merged**